File: n448-find-all-numbers-disappeared-in-an-array.js
var findDisappearedNumbers = function(nums) {
    nums.forEach((value, index) => {
        newIndex = Math.abs(value) - 1
        nums[newIndex] = -Math.abs(nums[newIndex])
    })
    
    console.log(nums)

    var ret = []
    nums.forEach((value, index) => {
        if (value > 0) {
            ret.push(index + 1)
        }
    })

    return ret
}


var a = [4,3,2,7,8,2,3,1]

console.log(findDisappearedNumbers(a))

File: n448-find-all-numbers-disappeared-in-an-array.go
package main

import "math"

func findDisappearedNumbers(nums []int) []int {
	// 遍历数组，以每个元素的绝对值减去1作为索引, 将数组的该索引位置元素置为负数
	// 之后再次遍历该数组，过滤出元素值依然为正数的数组索引，那么该索引位置+1即为源数组中没有出现过的数字
	for i := 0; i < len(nums); i++ {
		ele := nums[i]
		index := int(math.Abs(float64(ele)) - 1)
		nums[index] = -int(math.Abs(float64(nums[index])))
	}

	ret := make([]int, 0)

	for i := 0; i < len(nums); i++ {
		if nums[i] > 0 {
			ret = append(ret, i+1)
		}
	}

	return ret
}
